{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### -------------------------------------------------------------------------------- Glimmer --------------------------------------------------------------------------------\n",
    "\n",
    "> Installation: https://ccb.jhu.edu/software/glimmer/index.shtml"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Ecoli / Halo.. / Myco.. / Natro.. / Roseo... (Manually)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Step 1 of 4:  Finding long orfs for training\n",
      "Starting at Wed Apr 23 12:44:07 2025\n",
      "\n",
      "Sequence file = sequence.fasta\n",
      "Excluded regions file = none\n",
      "Circular genome = true\n",
      "Initial minimum gene length = 90 bp\n",
      "Determine optimal min gene length to maximize number of genes\n",
      "Maximum overlap bases = 30\n",
      "Start codons = atg,gtg,ttg\n",
      "Stop codons = taa,tag,tga\n",
      "Sequence length = 4641652\n",
      "Final minimum gene length = 463\n",
      "Number of genes = 2795\n",
      "Total bases = 3094821\n",
      "Step 2 of 4:  Extracting training sequences\n",
      "Step 3 of 4:  Building ICM\n",
      "Step 4 of 4:  Running Glimmer3\n",
      "Starting at Wed Apr 23 12:44:15 2025\n",
      "\n",
      "Sequence file = sequence.fasta\n",
      "Number of sequences = 1\n",
      "ICM model file = from-scratch.icm\n",
      "Excluded regions file = none\n",
      "List of orfs file = none\n",
      "Input is NOT separate orfs\n",
      "Independent (non-coding) scores are used\n",
      "Circular genome = true\n",
      "Truncated orfs = false\n",
      "Minimum gene length = 110 bp\n",
      "Maximum overlap bases = 50\n",
      "Threshold score = 30\n",
      "Use first start codon = false\n",
      "Start codons = atg,gtg,ttg\n",
      "Start probs = 0.600,0.300,0.100\n",
      "Stop codons = taa,tag,tga\n",
      "GC percentage = 50.8%\n",
      "Ignore score on orfs longer than 750\n",
      "Analyzing Sequence #1\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n"
     ]
    }
   ],
   "source": [
    "# Glimmer-from-scratch\n",
    "!cd Ecoli && csh g3-from-scratch.csh sequence.fasta  from-scratch\n",
    "!cd Ecoli && mkdir -p from-scratch\n",
    "!cd Ecoli && mv from-scratch.detail from-scratch.icm from-scratch.longorfs from-scratch.predict from-scratch.train from-scratch/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Step 1 of 8:  Finding long orfs for training\n",
      "Starting at Wed Apr 23 12:46:32 2025\n",
      "\n",
      "Sequence file = sequence.fasta\n",
      "Excluded regions file = none\n",
      "Circular genome = true\n",
      "Initial minimum gene length = 90 bp\n",
      "Determine optimal min gene length to maximize number of genes\n",
      "Maximum overlap bases = 30\n",
      "Start codons = atg,gtg,ttg\n",
      "Stop codons = taa,tag,tga\n",
      "Sequence length = 4641652\n",
      "Final minimum gene length = 463\n",
      "Number of genes = 2795\n",
      "Total bases = 3094821\n",
      "Step 2 of 8:  Extracting training sequences\n",
      "Step 3 of 8:  Building ICM\n",
      "Step 4 of 8:  Running first Glimmer3\n",
      "Starting at Wed Apr 23 12:46:41 2025\n",
      "\n",
      "Sequence file = sequence.fasta\n",
      "Number of sequences = 1\n",
      "ICM model file = iterated.icm\n",
      "Excluded regions file = none\n",
      "List of orfs file = none\n",
      "Input is NOT separate orfs\n",
      "Independent (non-coding) scores are used\n",
      "Circular genome = true\n",
      "Truncated orfs = false\n",
      "Minimum gene length = 110 bp\n",
      "Maximum overlap bases = 50\n",
      "Threshold score = 30\n",
      "Use first start codon = false\n",
      "Start codons = atg,gtg,ttg\n",
      "Start probs = 0.600,0.300,0.100\n",
      "Stop codons = taa,tag,tga\n",
      "GC percentage = 50.8%\n",
      "Ignore score on orfs longer than 750\n",
      "Analyzing Sequence #1\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n",
      "Step 5 of 8:  Getting training coordinates\n",
      "Step 6 of 8:  Making PWM from upstream regions\n",
      "[1]54: alignProb=11485.53 Info/param=0.06 diff=114.17    \n",
      "[2]48: alignProb=11408.61 Info/param=0.06 diff=67.50     \n",
      "[3]24: alignProb=11485.08 Info/param=0.06 diff=303.93    \n",
      "[4]26: alignProb=11321.50 Info/param=0.06 diff=141.63    \n",
      "[5]42: alignProb=11393.59 Info/param=0.06 diff=32.00     \n",
      "[6]19: alignProb=11347.91 Info/param=0.06 diff=179.08    \n",
      "[7]45: alignProb=11385.46 Info/param=0.06 diff=11.38     \n",
      "[8]48: alignProb=11371.78 Info/param=0.06 diff=5.49      \n",
      "[9]47: alignProb=11241.23 Info/param=0.06 diff=21.60     \n",
      "[10]33: alignProb=10864.03 Info/param=0.05 diff=9.33      \n",
      "Optimizing...\n",
      "New align prob=14637.34\n",
      "New align prob=14697.45\n",
      "New align prob=14716.25\n",
      "New align prob=14718.86\n",
      "New align prob=14720.22\n",
      "Step 7 of 8:  Getting start-codon usage\n",
      "Step 8 of 8:  Running second Glimmer3\n",
      "Starting at Wed Apr 23 12:47:11 2025\n",
      "\n",
      "Sequence file = sequence.fasta\n",
      "Number of sequences = 1\n",
      "ICM model file = iterated.icm\n",
      "Excluded regions file = none\n",
      "List of orfs file = none\n",
      "Input is NOT separate orfs\n",
      "Independent (non-coding) scores are used\n",
      "Circular genome = true\n",
      "Truncated orfs = false\n",
      "Minimum gene length = 110 bp\n",
      "Maximum overlap bases = 50\n",
      "Threshold score = 30\n",
      "Use first start codon = false\n",
      "Start codons = atg,gtg,ttg\n",
      "Start probs = 0.809,0.140,0.051\n",
      "Stop codons = taa,tag,tga\n",
      "GC percentage = 50.8%\n",
      "Ignore score on orfs longer than 750\n",
      "Analyzing Sequence #1\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n"
     ]
    }
   ],
   "source": [
    "# Glimmer-iterated\n",
    "!cd Ecoli && csh g3-iterated.csh sequence.fasta iterated\n",
    "!cd Ecoli && mkdir -p iterated\n",
    "!cd Ecoli && mv iterated.coords iterated.longorfs iterated.run1.detail iterated.upstream iterated.detail iterated.motif iterated.run1.predict iterated.icm iterated.predict iterated.train iterated/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "\n",
    "def process_gff(gff_path: str, output_path: str, portion: str = \"3.3\"):\n",
    "    gff_path = Path(gff_path)\n",
    "    output_path = Path(output_path)\n",
    "\n",
    "    df = pd.read_csv(gff_path, sep=\"\\t\", comment=\"#\", header=None)\n",
    "    df = df.drop(columns=[1, 2])\n",
    "    \n",
    "    total = len(df)\n",
    "    if portion == \"1.3\":\n",
    "        df = df.iloc[: (1 * total) // 3]\n",
    "    elif portion == \"1.2\":\n",
    "        df = df.iloc[: total // 2]\n",
    "    elif portion == \"2.3\":\n",
    "        df = df.iloc[: (2 * total) // 3]\n",
    "\n",
    "    df.to_csv(output_path, sep=\"\\t\", index=False, header=False)\n",
    "    print(f\"✅ Saved {portion} GFF to: {output_path}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Ecoli"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Saved 1.3 GFF to: Ecoli/coords-1.3.nh\n",
      "Step 1 of 5:  Extracting training sequences\n",
      "Step 2 of 5:  Building ICM\n",
      "Step 3 of 5:  Making PWM from upstream regions\n",
      "[1]39: alignProb=646.54 Info/param=0.05 diff=9.23     \n",
      "[2]35: alignProb=680.03 Info/param=0.05 diff=14.55     \n",
      "[3]21: alignProb=636.09 Info/param=0.05 diff=12.54    \n",
      "[4]56: alignProb=658.05 Info/param=0.05 diff=11.94    \n",
      "[5]15: alignProb=676.54 Info/param=0.05 diff=6.28     \n",
      "[6]46: alignProb=661.84 Info/param=0.05 diff=4.94     \n",
      "[7]40: alignProb=681.31 Info/param=0.05 diff=16.32    \n",
      "[8]26: alignProb=642.74 Info/param=0.05 diff=30.49    \n",
      "[9]23: alignProb=620.49 Info/param=0.05 diff=11.20    \n",
      "[10]44: alignProb=652.84 Info/param=0.05 diff=32.82    \n",
      "Optimizing...\n",
      "New align prob=803.92\n",
      "New align prob=806.45\n",
      "New align prob=806.35\n",
      "Step 4 of 5:  Getting start-codon usage\n",
      "Step 5 of 5:  Running Glimmer3\n",
      "Starting at Wed Apr 23 13:16:04 2025\n",
      "\n",
      "Sequence file = sequence.fasta\n",
      "Number of sequences = 1\n",
      "ICM model file = training.icm\n",
      "Excluded regions file = none\n",
      "List of orfs file = none\n",
      "Input is NOT separate orfs\n",
      "Independent (non-coding) scores are used\n",
      "Circular genome = true\n",
      "Truncated orfs = false\n",
      "Minimum gene length = 110 bp\n",
      "Maximum overlap bases = 50\n",
      "Threshold score = 30\n",
      "Use first start codon = false\n",
      "Start codons = atg,gtg,ttg\n",
      "Start probs = 0.480,0.051,0.012\n",
      "Stop codons = taa,tag,tga\n",
      "GC percentage = 50.8%\n",
      "Ignore score on orfs longer than 750\n",
      "Analyzing Sequence #1\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n"
     ]
    }
   ],
   "source": [
    "# Glimmer-from-training\n",
    "folder = \"Ecoli\"\n",
    "portion = \"1.3\"\n",
    "process_gff(\n",
    "    gff_path=f\"Verified-genome-ALL/{folder}/verified.gff\",\n",
    "    output_path=f\"{folder}/coords-{portion}.nh\",\n",
    "    portion=portion\n",
    ")\n",
    "\n",
    "# script\n",
    "!cd Ecoli && csh g3-from-training.csh sequence.fasta coords-1.3.nh training\n",
    "!cd Ecoli && mkdir -p training-1.3\n",
    "!cd Ecoli && mv training.detail training.icm training.motif training.predict training.train training.upstream training-1.3/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Saved 1.2 GFF to: Ecoli/coords-1.2.nh\n",
      "Step 1 of 5:  Extracting training sequences\n",
      "Step 2 of 5:  Building ICM\n",
      "Step 3 of 5:  Making PWM from upstream regions\n",
      "[1]33: alignProb=937.85 Info/param=0.05 diff=22.75    \n",
      "[2]37: alignProb=942.41 Info/param=0.04 diff=11.44    \n",
      "[3]30: alignProb=999.25 Info/param=0.06 diff=37.56    \n",
      "[4]42: alignProb=969.59 Info/param=0.05 diff=57.65    \n",
      "[5]17: alignProb=924.68 Info/param=0.05 diff=90.23    \n",
      "[6]18: alignProb=1012.51 Info/param=0.05 diff=55.05    \n",
      "[7]24: alignProb=950.57 Info/param=0.05 diff=44.89    \n",
      "[8]38: alignProb=919.25 Info/param=0.04 diff=12.02    \n",
      "[9]27: alignProb=931.07 Info/param=0.05 diff=41.04    \n",
      "[10]41: alignProb=947.13 Info/param=0.05 diff=28.49    \n",
      "Optimizing...\n",
      "New align prob=1212.50\n",
      "New align prob=1218.54\n",
      "New align prob=1219.95\n",
      "New align prob=1220.62\n",
      "Step 4 of 5:  Getting start-codon usage\n",
      "Step 5 of 5:  Running Glimmer3\n",
      "Starting at Wed Apr 23 13:16:23 2025\n",
      "\n",
      "Sequence file = sequence.fasta\n",
      "Number of sequences = 1\n",
      "ICM model file = training.icm\n",
      "Excluded regions file = none\n",
      "List of orfs file = none\n",
      "Input is NOT separate orfs\n",
      "Independent (non-coding) scores are used\n",
      "Circular genome = true\n",
      "Truncated orfs = false\n",
      "Minimum gene length = 110 bp\n",
      "Maximum overlap bases = 50\n",
      "Threshold score = 30\n",
      "Use first start codon = false\n",
      "Start codons = atg,gtg,ttg\n",
      "Start probs = 0.445,0.044,0.008\n",
      "Stop codons = taa,tag,tga\n",
      "GC percentage = 50.8%\n",
      "Ignore score on orfs longer than 750\n",
      "Analyzing Sequence #1\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n"
     ]
    }
   ],
   "source": [
    "# Glimmer-from-training\n",
    "folder = \"Ecoli\"\n",
    "portion = \"1.2\"\n",
    "process_gff(\n",
    "    gff_path=f\"Verified-genome-ALL/{folder}/verified.gff\",\n",
    "    output_path=f\"{folder}/coords-{portion}.nh\",\n",
    "    portion=portion\n",
    ")\n",
    "\n",
    "# script\n",
    "!cd Ecoli && csh g3-from-training.csh sequence.fasta coords-1.2.nh training\n",
    "!cd Ecoli && mkdir -p training-1.2\n",
    "!cd Ecoli && mv training.detail training.icm training.motif training.predict training.train training.upstream training-1.2/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Saved 2.3 GFF to: Ecoli/coords-2.3.nh\n",
      "Step 1 of 5:  Extracting training sequences\n",
      "Step 2 of 5:  Building ICM\n",
      "Step 3 of 5:  Making PWM from upstream regions\n",
      "[1]51: alignProb=1287.22 Info/param=0.05 diff=18.14    \n",
      "[2]34: alignProb=1135.11 Info/param=0.04 diff=27.94    \n",
      "[3]27: alignProb=1254.56 Info/param=0.05 diff=20.38    \n",
      "[4]65: alignProb=1175.11 Info/param=0.04 diff=4.11     \n",
      "[5]16: alignProb=1166.24 Info/param=0.04 diff=29.56    \n",
      "[6]28: alignProb=1138.63 Info/param=0.04 diff=30.89    \n",
      "[7]40: alignProb=1286.43 Info/param=0.05 diff=1.26     \n",
      "[8]49: alignProb=1251.24 Info/param=0.05 diff=10.49    \n",
      "[9]33: alignProb=1150.55 Info/param=0.04 diff=27.65    \n",
      "[10]23: alignProb=1283.72 Info/param=0.05 diff=18.83    \n",
      "Optimizing...\n",
      "New align prob=1600.34\n",
      "New align prob=1609.05\n",
      "New align prob=1612.15\n",
      "Step 4 of 5:  Getting start-codon usage\n",
      "Step 5 of 5:  Running Glimmer3\n",
      "Starting at Wed Apr 23 13:16:43 2025\n",
      "\n",
      "Sequence file = sequence.fasta\n",
      "Number of sequences = 1\n",
      "ICM model file = training.icm\n",
      "Excluded regions file = none\n",
      "List of orfs file = none\n",
      "Input is NOT separate orfs\n",
      "Independent (non-coding) scores are used\n",
      "Circular genome = true\n",
      "Truncated orfs = false\n",
      "Minimum gene length = 110 bp\n",
      "Maximum overlap bases = 50\n",
      "Threshold score = 30\n",
      "Use first start codon = false\n",
      "Start codons = atg,gtg,ttg\n",
      "Start probs = 0.418,0.037,0.008\n",
      "Stop codons = taa,tag,tga\n",
      "GC percentage = 50.8%\n",
      "Ignore score on orfs longer than 750\n",
      "Analyzing Sequence #1\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n"
     ]
    }
   ],
   "source": [
    "# Glimmer-from-training\n",
    "folder = \"Ecoli\"\n",
    "portion = \"2.3\"\n",
    "process_gff(\n",
    "    gff_path=f\"Verified-genome-ALL/{folder}/verified.gff\",\n",
    "    output_path=f\"{folder}/coords-{portion}.nh\",\n",
    "    portion=portion\n",
    ")\n",
    "\n",
    "# script\n",
    "!cd Ecoli && csh g3-from-training.csh sequence.fasta coords-2.3.nh training\n",
    "!cd Ecoli && mkdir -p training-2.3\n",
    "!cd Ecoli && mv training.detail training.icm training.motif training.predict training.train training.upstream training-2.3/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Saved 3.3 GFF to: Ecoli/coords-3.3.nh\n",
      "Step 1 of 5:  Extracting training sequences\n",
      "Step 2 of 5:  Building ICM\n",
      "Step 3 of 5:  Making PWM from upstream regions\n",
      "[1]56: alignProb=1859.65 Info/param=0.04 diff=38.47     \n",
      "[2]33: alignProb=1543.60 Info/param=0.04 diff=12.60     \n",
      "[3]57: alignProb=1630.46 Info/param=0.03 diff=3.16      \n",
      "[4]29: alignProb=1567.92 Info/param=0.04 diff=6.29      \n",
      "[5]49: alignProb=1622.05 Info/param=0.04 diff=40.74    \n",
      "[6]41: alignProb=1622.61 Info/param=0.04 diff=29.99     \n",
      "[7]48: alignProb=1829.97 Info/param=0.04 diff=50.67     \n",
      "[8]47: alignProb=1770.18 Info/param=0.04 diff=45.75     \n",
      "[9]52: alignProb=1379.91 Info/param=0.02 diff=6.91     \n",
      "[10]30: alignProb=1284.92 Info/param=0.03 diff=100.86    \n",
      "Optimizing...\n",
      "New align prob=2342.71\n",
      "New align prob=2352.49\n",
      "New align prob=2353.63\n",
      "New align prob=2355.09\n",
      "New align prob=2355.27\n",
      "New align prob=2355.30\n",
      "New align prob=2356.46\n",
      "Step 4 of 5:  Getting start-codon usage\n",
      "Step 5 of 5:  Running Glimmer3\n",
      "Starting at Wed Apr 23 13:17:04 2025\n",
      "\n",
      "Sequence file = sequence.fasta\n",
      "Number of sequences = 1\n",
      "ICM model file = training.icm\n",
      "Excluded regions file = none\n",
      "List of orfs file = none\n",
      "Input is NOT separate orfs\n",
      "Independent (non-coding) scores are used\n",
      "Circular genome = true\n",
      "Truncated orfs = false\n",
      "Minimum gene length = 110 bp\n",
      "Maximum overlap bases = 50\n",
      "Threshold score = 30\n",
      "Use first start codon = false\n",
      "Start codons = atg,gtg,ttg\n",
      "Start probs = 0.407,0.035,0.008\n",
      "Stop codons = taa,tag,tga\n",
      "GC percentage = 50.8%\n",
      "Ignore score on orfs longer than 750\n",
      "Analyzing Sequence #1\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n"
     ]
    }
   ],
   "source": [
    "# Glimmer-from-training\n",
    "folder = \"Ecoli\"\n",
    "portion = \"3.3\"\n",
    "process_gff(\n",
    "    gff_path=f\"Verified-genome-ALL/{folder}/verified.gff\",\n",
    "    output_path=f\"{folder}/coords-{portion}.nh\",\n",
    "    portion=portion\n",
    ")\n",
    "\n",
    "# script\n",
    "!cd Ecoli && csh g3-from-training.csh sequence.fasta coords-3.3.nh training\n",
    "!cd Ecoli && mkdir -p training-3.3\n",
    "!cd Ecoli && mv training.detail training.icm training.motif training.predict training.train training.upstream training-3.3/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Halobacterium"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Saved 1.3 GFF to: Halobacterium/coords-1.3.nh\n",
      "Step 1 of 5:  Extracting training sequences\n",
      "Step 2 of 5:  Building ICM\n",
      "Step 3 of 5:  Making PWM from upstream regions\n",
      "[1]69: alignProb=437.38 Info/param=0.03 diff=5.99     \n",
      "[2]24: alignProb=386.53 Info/param=0.03 diff=4.99     \n",
      "[3]38: alignProb=443.49 Info/param=0.03 diff=14.24    \n",
      "[4]28: alignProb=419.12 Info/param=0.02 diff=23.79    \n",
      "[5]42: alignProb=430.07 Info/param=0.03 diff=3.32     \n",
      "[6]24: alignProb=387.38 Info/param=0.03 diff=11.73    \n",
      "[7]110: alignProb=455.38 Info/param=0.04 diff=3.20    \n",
      "[8]21: alignProb=420.25 Info/param=0.03 diff=0.81     \n",
      "[9]42: alignProb=429.99 Info/param=0.03 diff=10.24    \n",
      "[10]39: alignProb=439.61 Info/param=0.04 diff=16.67    \n",
      "Optimizing...\n",
      "New align prob=507.28\n",
      "New align prob=510.89\n",
      "New align prob=511.21\n",
      "Step 4 of 5:  Getting start-codon usage\n",
      "Step 5 of 5:  Running Glimmer3\n",
      "Starting at Wed Apr 23 13:17:22 2025\n",
      "\n",
      "Sequence file = sequence.fasta\n",
      "Number of sequences = 5\n",
      "ICM model file = training.icm\n",
      "Excluded regions file = none\n",
      "List of orfs file = none\n",
      "Input is NOT separate orfs\n",
      "Independent (non-coding) scores are used\n",
      "Circular genome = true\n",
      "Truncated orfs = false\n",
      "Minimum gene length = 110 bp\n",
      "Maximum overlap bases = 50\n",
      "Threshold score = 30\n",
      "Use first start codon = false\n",
      "Start codons = atg,gtg,ttg\n",
      "Start probs = 0.403,0.040,0.000\n",
      "Stop codons = taa,tag,tga\n",
      "GC percentage = 65.7%\n",
      "Ignore score on orfs longer than 1329\n",
      "Analyzing Sequence #1\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n",
      "Analyzing Sequence #2\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n",
      "Analyzing Sequence #3\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n",
      "Analyzing Sequence #4\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n",
      "Analyzing Sequence #5\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n"
     ]
    }
   ],
   "source": [
    "# Glimmer-from-training\n",
    "folder = \"Halobacterium\"\n",
    "portion = \"1.3\"\n",
    "process_gff(\n",
    "    gff_path=f\"Verified-genome-ALL/{folder}/verified.gff\",\n",
    "    output_path=f\"{folder}/coords-{portion}.nh\",\n",
    "    portion=portion\n",
    ")\n",
    "\n",
    "# script\n",
    "!cd Halobacterium && csh g3-from-training.csh sequence.fasta coords-1.3.nh training\n",
    "!cd Halobacterium && mkdir -p training-1.3\n",
    "!cd Halobacterium && mv training.detail training.icm training.motif training.predict training.train training.upstream training-1.3/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Saved 1.2 GFF to: Halobacterium/coords-1.2.nh\n",
      "Step 1 of 5:  Extracting training sequences\n",
      "Step 2 of 5:  Building ICM\n",
      "Step 3 of 5:  Making PWM from upstream regions\n",
      "[1]76: alignProb=645.07 Info/param=0.03 diff=36.15    \n",
      "[2]43: alignProb=586.16 Info/param=0.02 diff=11.45    \n",
      "[3]46: alignProb=605.81 Info/param=0.03 diff=20.31    \n",
      "[4]15: alignProb=526.23 Info/param=0.02 diff=25.99    \n",
      "[5]43: alignProb=607.41 Info/param=0.03 diff=15.23    \n",
      "[6]30: alignProb=544.83 Info/param=0.02 diff=4.01     \n",
      "[7]36: alignProb=622.22 Info/param=0.03 diff=6.49     \n",
      "[8]49: alignProb=586.05 Info/param=0.02 diff=3.89     \n",
      "[9]30: alignProb=552.42 Info/param=0.02 diff=12.46    \n",
      "[10]23: alignProb=565.64 Info/param=0.03 diff=18.20    \n",
      "Optimizing...\n",
      "New align prob=738.31\n",
      "New align prob=742.22\n",
      "New align prob=742.61\n",
      "Step 4 of 5:  Getting start-codon usage\n",
      "Step 5 of 5:  Running Glimmer3\n",
      "Starting at Wed Apr 23 13:17:36 2025\n",
      "\n",
      "Sequence file = sequence.fasta\n",
      "Number of sequences = 5\n",
      "ICM model file = training.icm\n",
      "Excluded regions file = none\n",
      "List of orfs file = none\n",
      "Input is NOT separate orfs\n",
      "Independent (non-coding) scores are used\n",
      "Circular genome = true\n",
      "Truncated orfs = false\n",
      "Minimum gene length = 110 bp\n",
      "Maximum overlap bases = 50\n",
      "Threshold score = 30\n",
      "Use first start codon = false\n",
      "Start codons = atg,gtg,ttg\n",
      "Start probs = 0.426,0.030,0.000\n",
      "Stop codons = taa,tag,tga\n",
      "GC percentage = 65.7%\n",
      "Ignore score on orfs longer than 1329\n",
      "Analyzing Sequence #1\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n",
      "Analyzing Sequence #2\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n",
      "Analyzing Sequence #3\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n",
      "Analyzing Sequence #4\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n",
      "Analyzing Sequence #5\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n"
     ]
    }
   ],
   "source": [
    "# Glimmer-from-training\n",
    "folder = \"Halobacterium\"\n",
    "portion = \"1.2\"\n",
    "process_gff(\n",
    "    gff_path=f\"Verified-genome-ALL/{folder}/verified.gff\",\n",
    "    output_path=f\"{folder}/coords-{portion}.nh\",\n",
    "    portion=portion\n",
    ")\n",
    "\n",
    "# script\n",
    "!cd Halobacterium && csh g3-from-training.csh sequence.fasta coords-1.2.nh training\n",
    "!cd Halobacterium && mkdir -p training-1.2\n",
    "!cd Halobacterium && mv training.detail training.icm training.motif training.predict training.train training.upstream training-1.2/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Saved 2.3 GFF to: Halobacterium/coords-2.3.nh\n",
      "Step 1 of 5:  Extracting training sequences\n",
      "Step 2 of 5:  Building ICM\n",
      "Step 3 of 5:  Making PWM from upstream regions\n",
      "[1]58: alignProb=752.39 Info/param=0.02 diff=9.02     \n",
      "[2]15: alignProb=659.93 Info/param=0.03 diff=33.28    \n",
      "[3]46: alignProb=709.59 Info/param=0.02 diff=4.90     \n",
      "[4]48: alignProb=701.08 Info/param=0.02 diff=24.02    \n",
      "[5]53: alignProb=702.34 Info/param=0.03 diff=18.47    \n",
      "[6]32: alignProb=737.40 Info/param=0.02 diff=42.51    \n",
      "[7]24: alignProb=712.43 Info/param=0.02 diff=48.73     \n",
      "[8]61: alignProb=798.34 Info/param=0.03 diff=6.88     \n",
      "[9]57: alignProb=810.66 Info/param=0.03 diff=1.45      \n",
      "[10]24: alignProb=670.79 Info/param=0.02 diff=2.32     \n",
      "Optimizing...\n",
      "New align prob=969.25\n",
      "New align prob=973.60\n",
      "New align prob=974.33\n",
      "Step 4 of 5:  Getting start-codon usage\n",
      "Step 5 of 5:  Running Glimmer3\n",
      "Starting at Wed Apr 23 13:17:51 2025\n",
      "\n",
      "Sequence file = sequence.fasta\n",
      "Number of sequences = 5\n",
      "ICM model file = training.icm\n",
      "Excluded regions file = none\n",
      "List of orfs file = none\n",
      "Input is NOT separate orfs\n",
      "Independent (non-coding) scores are used\n",
      "Circular genome = true\n",
      "Truncated orfs = false\n",
      "Minimum gene length = 110 bp\n",
      "Maximum overlap bases = 50\n",
      "Threshold score = 30\n",
      "Use first start codon = false\n",
      "Start codons = atg,gtg,ttg\n",
      "Start probs = 0.465,0.028,0.000\n",
      "Stop codons = taa,tag,tga\n",
      "GC percentage = 65.7%\n",
      "Ignore score on orfs longer than 1329\n",
      "Analyzing Sequence #1\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n",
      "Analyzing Sequence #2\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n",
      "Analyzing Sequence #3\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n",
      "Analyzing Sequence #4\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n",
      "Analyzing Sequence #5\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n"
     ]
    }
   ],
   "source": [
    "# Glimmer-from-training\n",
    "folder = \"Halobacterium\"\n",
    "portion = \"2.3\"\n",
    "process_gff(\n",
    "    gff_path=f\"Verified-genome-ALL/{folder}/verified.gff\",\n",
    "    output_path=f\"{folder}/coords-{portion}.nh\",\n",
    "    portion=portion\n",
    ")\n",
    "\n",
    "# script\n",
    "!cd Halobacterium && csh g3-from-training.csh sequence.fasta coords-2.3.nh training\n",
    "!cd Halobacterium && mkdir -p training-2.3\n",
    "!cd Halobacterium && mv training.detail training.icm training.motif training.predict training.train training.upstream training-2.3/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Saved 3.3 GFF to: Halobacterium/coords-3.3.nh\n",
      "Step 1 of 5:  Extracting training sequences\n",
      "Step 2 of 5:  Building ICM\n",
      "Step 3 of 5:  Making PWM from upstream regions\n",
      "[1]31: alignProb=1007.76 Info/param=0.02 diff=7.62     \n",
      "[2]58: alignProb=1087.23 Info/param=0.02 diff=18.21    \n",
      "[3]18: alignProb=957.44 Info/param=0.02 diff=1.68     \n",
      "[4]61: alignProb=1078.81 Info/param=0.02 diff=39.10    \n",
      "[5]50: alignProb=1019.84 Info/param=0.02 diff=10.28    \n",
      "[6]41: alignProb=996.23 Info/param=0.02 diff=4.20     \n",
      "[7]19: alignProb=989.31 Info/param=0.02 diff=7.99      \n",
      "[8]42: alignProb=1061.92 Info/param=0.02 diff=77.04    \n",
      "[9]49: alignProb=1051.23 Info/param=0.02 diff=5.31     \n",
      "[10]28: alignProb=951.01 Info/param=0.02 diff=0.49      \n",
      "Optimizing...\n",
      "New align prob=1391.58\n",
      "New align prob=1398.76\n",
      "New align prob=1402.70\n",
      "New align prob=1405.36\n",
      "New align prob=1408.24\n",
      "New align prob=1409.40\n",
      "New align prob=1409.50\n",
      "Step 4 of 5:  Getting start-codon usage\n",
      "Step 5 of 5:  Running Glimmer3\n",
      "Starting at Wed Apr 23 13:18:07 2025\n",
      "\n",
      "Sequence file = sequence.fasta\n",
      "Number of sequences = 5\n",
      "ICM model file = training.icm\n",
      "Excluded regions file = none\n",
      "List of orfs file = none\n",
      "Input is NOT separate orfs\n",
      "Independent (non-coding) scores are used\n",
      "Circular genome = true\n",
      "Truncated orfs = false\n",
      "Minimum gene length = 110 bp\n",
      "Maximum overlap bases = 50\n",
      "Threshold score = 30\n",
      "Use first start codon = false\n",
      "Start codons = atg,gtg,ttg\n",
      "Start probs = 0.457,0.025,0.000\n",
      "Stop codons = taa,tag,tga\n",
      "GC percentage = 65.7%\n",
      "Ignore score on orfs longer than 1329\n",
      "Analyzing Sequence #1\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n",
      "Analyzing Sequence #2\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n",
      "Analyzing Sequence #3\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n",
      "Analyzing Sequence #4\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n",
      "Analyzing Sequence #5\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n"
     ]
    }
   ],
   "source": [
    "# Glimmer-from-training\n",
    "folder = \"Halobacterium\"\n",
    "portion = \"3.3\"\n",
    "process_gff(\n",
    "    gff_path=f\"Verified-genome-ALL/{folder}/verified.gff\",\n",
    "    output_path=f\"{folder}/coords-{portion}.nh\",\n",
    "    portion=portion\n",
    ")\n",
    "\n",
    "# script\n",
    "!cd Halobacterium && csh g3-from-training.csh sequence.fasta coords-3.3.nh training\n",
    "!cd Halobacterium && mkdir -p training-3.3\n",
    "!cd Halobacterium && mv training.detail training.icm training.motif training.predict training.train training.upstream training-3.3/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Mycobacterium"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Saved 1.3 GFF to: Mycobacterium/coords-1.3.nh\n",
      "Step 1 of 5:  Extracting training sequences\n",
      "Step 2 of 5:  Building ICM\n",
      "Step 3 of 5:  Making PWM from upstream regions\n",
      "[1]72: alignProb=491.46 Info/param=0.01 diff=4.65     \n",
      "[2]59: alignProb=517.74 Info/param=0.02 diff=0.40     \n",
      "[3]63: alignProb=475.25 Info/param=0.02 diff=27.62    \n",
      "[4]20: alignProb=477.64 Info/param=0.02 diff=15.89    \n",
      "[5]59: alignProb=502.05 Info/param=0.02 diff=8.98     \n",
      "[6]83: alignProb=515.07 Info/param=0.02 diff=6.06     \n",
      "[7]45: alignProb=544.40 Info/param=0.02 diff=18.60    \n",
      "[8]50: alignProb=517.28 Info/param=0.02 diff=25.59    \n",
      "[9]71: alignProb=469.45 Info/param=0.02 diff=25.98    \n",
      "[10]37: alignProb=511.32 Info/param=0.02 diff=15.74    \n",
      "Optimizing...\n",
      "New align prob=643.26\n",
      "New align prob=647.38\n",
      "New align prob=648.08\n",
      "New align prob=648.55\n",
      "Step 4 of 5:  Getting start-codon usage\n",
      "Step 5 of 5:  Running Glimmer3\n",
      "Starting at Wed Apr 23 16:38:32 2025\n",
      "\n",
      "Sequence file = sequence.fasta\n",
      "Number of sequences = 1\n",
      "ICM model file = training.icm\n",
      "Excluded regions file = none\n",
      "List of orfs file = none\n",
      "Input is NOT separate orfs\n",
      "Independent (non-coding) scores are used\n",
      "Circular genome = true\n",
      "Truncated orfs = false\n",
      "Minimum gene length = 110 bp\n",
      "Maximum overlap bases = 50\n",
      "Threshold score = 30\n",
      "Use first start codon = false\n",
      "Start codons = atg,gtg,ttg\n",
      "Start probs = 0.348,0.172,0.013\n",
      "Stop codons = taa,tag,tga\n",
      "GC percentage = 65.6%\n",
      "Ignore score on orfs longer than 1323\n",
      "Analyzing Sequence #1\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n"
     ]
    }
   ],
   "source": [
    "# Glimmer-from-training\n",
    "folder = \"Mycobacterium\"\n",
    "portion = \"1.3\"\n",
    "process_gff(\n",
    "    gff_path=f\"Verified-genome-ALL/{folder}/verified.gff\",\n",
    "    output_path=f\"{folder}/coords-{portion}.nh\",\n",
    "    portion=portion\n",
    ")\n",
    "\n",
    "# script\n",
    "!cd Mycobacterium && csh g3-from-training.csh sequence.fasta coords-1.3.nh training\n",
    "!cd Mycobacterium && mkdir -p training-1.3\n",
    "!cd Mycobacterium && mv training.detail training.icm training.motif training.predict training.train training.upstream training-1.3/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Saved 1.2 GFF to: Mycobacterium/coords-1.2.nh\n",
      "Step 1 of 5:  Extracting training sequences\n",
      "Step 2 of 5:  Building ICM\n",
      "Step 3 of 5:  Making PWM from upstream regions\n",
      "[1]58: alignProb=711.55 Info/param=0.02 diff=1.82     \n",
      "[2]46: alignProb=668.50 Info/param=0.02 diff=16.42    \n",
      "[3]38: alignProb=630.56 Info/param=0.02 diff=49.27    \n",
      "[4]44: alignProb=680.60 Info/param=0.02 diff=13.35    \n",
      "[5]35: alignProb=646.38 Info/param=0.02 diff=42.64    \n",
      "[6]34: alignProb=722.46 Info/param=0.02 diff=67.93    \n",
      "[7]25: alignProb=618.21 Info/param=0.02 diff=21.31    \n",
      "[8]62: alignProb=675.70 Info/param=0.02 diff=4.53     \n",
      "[9]36: alignProb=662.46 Info/param=0.02 diff=25.77    \n",
      "[10]33: alignProb=686.35 Info/param=0.02 diff=2.07     \n",
      "Optimizing...\n",
      "New align prob=918.30\n",
      "New align prob=921.57\n",
      "New align prob=923.39\n",
      "New align prob=926.21\n",
      "New align prob=928.82\n",
      "New align prob=931.62\n",
      "New align prob=932.77\n",
      "New align prob=933.64\n",
      "Step 4 of 5:  Getting start-codon usage\n",
      "Step 5 of 5:  Running Glimmer3\n",
      "Starting at Wed Apr 23 16:38:02 2025\n",
      "\n",
      "Sequence file = sequence.fasta\n",
      "Number of sequences = 1\n",
      "ICM model file = training.icm\n",
      "Excluded regions file = none\n",
      "List of orfs file = none\n",
      "Input is NOT separate orfs\n",
      "Independent (non-coding) scores are used\n",
      "Circular genome = true\n",
      "Truncated orfs = false\n",
      "Minimum gene length = 110 bp\n",
      "Maximum overlap bases = 50\n",
      "Threshold score = 30\n",
      "Use first start codon = false\n",
      "Start codons = atg,gtg,ttg\n",
      "Start probs = 0.403,0.183,0.017\n",
      "Stop codons = taa,tag,tga\n",
      "GC percentage = 65.6%\n",
      "Ignore score on orfs longer than 1323\n",
      "Analyzing Sequence #1\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n"
     ]
    }
   ],
   "source": [
    "# Glimmer-from-training\n",
    "folder = \"Mycobacterium\"\n",
    "portion = \"1.2\"\n",
    "process_gff(\n",
    "    gff_path=f\"Verified-genome-ALL/{folder}/verified.gff\",\n",
    "    output_path=f\"{folder}/coords-{portion}.nh\",\n",
    "    portion=portion\n",
    ")\n",
    "\n",
    "# script\n",
    "!cd Mycobacterium && csh g3-from-training.csh sequence.fasta coords-1.2.nh training\n",
    "!cd Mycobacterium && mkdir -p training-1.2\n",
    "!cd Mycobacterium && mv training.detail training.icm training.motif training.predict training.train training.upstream training-1.2/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Saved 2.3 GFF to: Mycobacterium/coords-2.3.nh\n",
      "Step 1 of 5:  Extracting training sequences\n",
      "Step 2 of 5:  Building ICM\n",
      "Step 3 of 5:  Making PWM from upstream regions\n",
      "[1]56: alignProb=898.79 Info/param=0.01 diff=12.01    \n",
      "[2]48: alignProb=825.22 Info/param=0.02 diff=19.15    \n",
      "[3]43: alignProb=764.30 Info/param=0.02 diff=37.25    \n",
      "[4]60: alignProb=884.42 Info/param=0.02 diff=3.29     \n",
      "[5]41: alignProb=813.71 Info/param=0.01 diff=20.10     \n",
      "[6]87: alignProb=919.83 Info/param=0.02 diff=13.72     \n",
      "[7]42: alignProb=868.47 Info/param=0.02 diff=47.05     \n",
      "[8]89: alignProb=864.76 Info/param=0.01 diff=3.89     \n",
      "[9]58: alignProb=821.65 Info/param=0.02 diff=32.97    \n",
      "[10]36: alignProb=803.35 Info/param=0.02 diff=14.14    \n",
      "Optimizing...\n",
      "New align prob=1229.15\n",
      "New align prob=1232.70\n",
      "New align prob=1233.05\n",
      "Step 4 of 5:  Getting start-codon usage\n",
      "Step 5 of 5:  Running Glimmer3\n",
      "Starting at Wed Apr 23 13:19:09 2025\n",
      "\n",
      "Sequence file = sequence.fasta\n",
      "Number of sequences = 1\n",
      "ICM model file = training.icm\n",
      "Excluded regions file = none\n",
      "List of orfs file = none\n",
      "Input is NOT separate orfs\n",
      "Independent (non-coding) scores are used\n",
      "Circular genome = true\n",
      "Truncated orfs = false\n",
      "Minimum gene length = 110 bp\n",
      "Maximum overlap bases = 50\n",
      "Threshold score = 30\n",
      "Use first start codon = false\n",
      "Start codons = atg,gtg,ttg\n",
      "Start probs = 0.347,0.180,0.013\n",
      "Stop codons = taa,tag,tga\n",
      "GC percentage = 65.6%\n",
      "Ignore score on orfs longer than 1323\n",
      "Analyzing Sequence #1\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n"
     ]
    }
   ],
   "source": [
    "# Glimmer-from-training\n",
    "folder = \"Mycobacterium\"\n",
    "portion = \"2.3\"\n",
    "process_gff(\n",
    "    gff_path=f\"Verified-genome-ALL/{folder}/verified.gff\",\n",
    "    output_path=f\"{folder}/coords-{portion}.nh\",\n",
    "    portion=portion\n",
    ")\n",
    "\n",
    "# script\n",
    "!cd Mycobacterium && csh g3-from-training.csh sequence.fasta coords-2.3.nh training\n",
    "!cd Mycobacterium && mkdir -p training-2.3\n",
    "!cd Mycobacterium && mv training.detail training.icm training.motif training.predict training.train training.upstream training-2.3/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Saved 3.3 GFF to: Mycobacterium/coords-3.3.nh\n",
      "Step 1 of 5:  Extracting training sequences\n",
      "Step 2 of 5:  Building ICM\n",
      "Step 3 of 5:  Making PWM from upstream regions\n",
      "[1]28: alignProb=1053.70 Info/param=0.02 diff=1.12      \n",
      "[2]33: alignProb=1018.16 Info/param=0.01 diff=15.44     \n",
      "[3]74: alignProb=1129.16 Info/param=0.02 diff=5.33     \n",
      "[4]40: alignProb=1095.45 Info/param=0.01 diff=15.55    \n",
      "[5]32: alignProb=1009.71 Info/param=0.01 diff=15.55    \n",
      "[6]35: alignProb=950.86 Info/param=0.01 diff=6.22      \n",
      "[7]39: alignProb=1041.56 Info/param=0.01 diff=42.11    \n",
      "[8]57: alignProb=994.57 Info/param=0.01 diff=10.24    \n",
      "[9]22: alignProb=1029.65 Info/param=0.02 diff=5.61     \n",
      "[10]34: alignProb=1122.30 Info/param=0.01 diff=6.04     \n",
      "Optimizing...\n",
      "New align prob=1751.25\n",
      "New align prob=1783.92\n",
      "New align prob=1805.86\n",
      "New align prob=1809.08\n",
      "New align prob=1809.64\n",
      "Step 4 of 5:  Getting start-codon usage\n",
      "Step 5 of 5:  Running Glimmer3\n",
      "Starting at Wed Apr 23 13:19:34 2025\n",
      "\n",
      "Sequence file = sequence.fasta\n",
      "Number of sequences = 1\n",
      "ICM model file = training.icm\n",
      "Excluded regions file = none\n",
      "List of orfs file = none\n",
      "Input is NOT separate orfs\n",
      "Independent (non-coding) scores are used\n",
      "Circular genome = true\n",
      "Truncated orfs = false\n",
      "Minimum gene length = 110 bp\n",
      "Maximum overlap bases = 50\n",
      "Threshold score = 30\n",
      "Use first start codon = false\n",
      "Start codons = atg,gtg,ttg\n",
      "Start probs = 0.327,0.155,0.010\n",
      "Stop codons = taa,tag,tga\n",
      "GC percentage = 65.6%\n",
      "Ignore score on orfs longer than 1323\n",
      "Analyzing Sequence #1\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n"
     ]
    }
   ],
   "source": [
    "# Glimmer-from-training\n",
    "folder = \"Mycobacterium\"\n",
    "portion = \"3.3\"\n",
    "process_gff(\n",
    "    gff_path=f\"Verified-genome-ALL/{folder}/verified.gff\",\n",
    "    output_path=f\"{folder}/coords-{portion}.nh\",\n",
    "    portion=portion\n",
    ")\n",
    "\n",
    "# script\n",
    "!cd Mycobacterium && csh g3-from-training.csh sequence.fasta coords-3.3.nh training\n",
    "!cd Mycobacterium && mkdir -p training-3.3\n",
    "!cd Mycobacterium && mv training.detail training.icm training.motif training.predict training.train training.upstream training-3.3/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Natronomonas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Saved 1.3 GFF to: Natronomonas/coords-1.3.nh\n",
      "Step 1 of 5:  Extracting training sequences\n",
      "Step 2 of 5:  Building ICM\n",
      "Step 3 of 5:  Making PWM from upstream regions\n",
      "[1]39: alignProb=244.05 Info/param=0.03 diff=7.65     \n",
      "[2]24: alignProb=251.42 Info/param=0.03 diff=1.24     \n",
      "[3]37: alignProb=262.66 Info/param=0.04 diff=5.29     \n",
      "[4]11: alignProb=283.39 Info/param=0.04 diff=36.40    \n",
      "[5]40: alignProb=283.67 Info/param=0.04 diff=4.54     \n",
      "[6]18: alignProb=254.42 Info/param=0.03 diff=22.76    \n",
      "[7]7: alignProb=263.57 Info/param=0.04 diff=8.86     \n",
      "[8]41: alignProb=269.98 Info/param=0.03 diff=2.84     \n",
      "[9]24: alignProb=278.13 Info/param=0.04 diff=16.73    \n",
      "[10]91: alignProb=276.59 Info/param=0.03 diff=3.51     \n",
      "Optimizing...\n",
      "New align prob=310.44\n",
      "Step 4 of 5:  Getting start-codon usage\n",
      "Step 5 of 5:  Running Glimmer3\n",
      "Starting at Wed Apr 23 13:19:55 2025\n",
      "\n",
      "Sequence file = sequence.fasta\n",
      "Number of sequences = 3\n",
      "ICM model file = training.icm\n",
      "Excluded regions file = none\n",
      "List of orfs file = none\n",
      "Input is NOT separate orfs\n",
      "Independent (non-coding) scores are used\n",
      "Circular genome = true\n",
      "Truncated orfs = false\n",
      "Minimum gene length = 110 bp\n",
      "Maximum overlap bases = 50\n",
      "Threshold score = 30\n",
      "Use first start codon = false\n",
      "Start codons = atg,gtg,ttg\n",
      "Start probs = 0.590,0.010,0.000\n",
      "Stop codons = taa,tag,tga\n",
      "GC percentage = 63.1%\n",
      "Ignore score on orfs longer than 1181\n",
      "Analyzing Sequence #1\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n",
      "Analyzing Sequence #2\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n",
      "Analyzing Sequence #3\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n"
     ]
    }
   ],
   "source": [
    "# Glimmer-from-training\n",
    "folder = \"Natronomonas\"\n",
    "portion = \"1.3\"\n",
    "process_gff(\n",
    "    gff_path=f\"Verified-genome-ALL/{folder}/verified.gff\",\n",
    "    output_path=f\"{folder}/coords-{portion}.nh\",\n",
    "    portion=portion\n",
    ")\n",
    "\n",
    "# script\n",
    "!cd Natronomonas && csh g3-from-training.csh sequence.fasta coords-1.3.nh training\n",
    "!cd Natronomonas && mkdir -p training-1.3\n",
    "!cd Natronomonas && mv training.detail training.icm training.motif training.predict training.train training.upstream training-1.3/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Saved 1.2 GFF to: Natronomonas/coords-1.2.nh\n",
      "Step 1 of 5:  Extracting training sequences\n",
      "Step 2 of 5:  Building ICM\n",
      "Step 3 of 5:  Making PWM from upstream regions\n",
      "[1]73: alignProb=389.29 Info/param=0.04 diff=1.95     \n",
      "[2]34: alignProb=399.73 Info/param=0.04 diff=6.70     \n",
      "[3]54: alignProb=393.43 Info/param=0.03 diff=3.96     \n",
      "[4]11: alignProb=360.35 Info/param=0.03 diff=43.70    \n",
      "[5]21: alignProb=398.10 Info/param=0.04 diff=0.87     \n",
      "[6]51: alignProb=398.43 Info/param=0.03 diff=11.91    \n",
      "[7]40: alignProb=350.25 Info/param=0.03 diff=16.41    \n",
      "[8]47: alignProb=404.71 Info/param=0.03 diff=12.72    \n",
      "[9]32: alignProb=397.48 Info/param=0.03 diff=4.10     \n",
      "[10]48: alignProb=359.70 Info/param=0.03 diff=4.24     \n",
      "Optimizing...\n",
      "New align prob=448.36\n",
      "New align prob=451.19\n",
      "New align prob=452.74\n",
      "New align prob=453.72\n",
      "New align prob=454.51\n",
      "New align prob=455.13\n",
      "Step 4 of 5:  Getting start-codon usage\n",
      "Step 5 of 5:  Running Glimmer3\n",
      "Starting at Wed Apr 23 13:20:09 2025\n",
      "\n",
      "Sequence file = sequence.fasta\n",
      "Number of sequences = 3\n",
      "ICM model file = training.icm\n",
      "Excluded regions file = none\n",
      "List of orfs file = none\n",
      "Input is NOT separate orfs\n",
      "Independent (non-coding) scores are used\n",
      "Circular genome = true\n",
      "Truncated orfs = false\n",
      "Minimum gene length = 110 bp\n",
      "Maximum overlap bases = 50\n",
      "Threshold score = 30\n",
      "Use first start codon = false\n",
      "Start codons = atg,gtg,ttg\n",
      "Start probs = 0.580,0.006,0.006\n",
      "Stop codons = taa,tag,tga\n",
      "GC percentage = 63.1%\n",
      "Ignore score on orfs longer than 1181\n",
      "Analyzing Sequence #1\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n",
      "Analyzing Sequence #2\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n",
      "Analyzing Sequence #3\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n"
     ]
    }
   ],
   "source": [
    "# Glimmer-from-training\n",
    "folder = \"Natronomonas\"\n",
    "portion = \"1.2\"\n",
    "process_gff(\n",
    "    gff_path=f\"Verified-genome-ALL/{folder}/verified.gff\",\n",
    "    output_path=f\"{folder}/coords-{portion}.nh\",\n",
    "    portion=portion\n",
    ")\n",
    "\n",
    "# script\n",
    "!cd Natronomonas && csh g3-from-training.csh sequence.fasta coords-1.2.nh training\n",
    "!cd Natronomonas && mkdir -p training-1.2\n",
    "!cd Natronomonas && mv training.detail training.icm training.motif training.predict training.train training.upstream training-1.2/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Saved 2.3 GFF to: Natronomonas/coords-2.3.nh\n",
      "Step 1 of 5:  Extracting training sequences\n",
      "Step 2 of 5:  Building ICM\n",
      "Step 3 of 5:  Making PWM from upstream regions\n",
      "[1]47: alignProb=532.12 Info/param=0.03 diff=3.22     \n",
      "[2]27: alignProb=577.55 Info/param=0.04 diff=32.93    \n",
      "[3]44: alignProb=531.01 Info/param=0.03 diff=14.69    \n",
      "[4]28: alignProb=532.08 Info/param=0.03 diff=6.60     \n",
      "[5]44: alignProb=527.05 Info/param=0.03 diff=0.75     \n",
      "[6]53: alignProb=521.06 Info/param=0.04 diff=38.79    \n",
      "[7]46: alignProb=542.59 Info/param=0.03 diff=6.30     \n",
      "[8]28: alignProb=477.61 Info/param=0.03 diff=27.72    \n",
      "[9]24: alignProb=528.62 Info/param=0.04 diff=10.86    \n",
      "[10]33: alignProb=524.78 Info/param=0.04 diff=1.08     \n",
      "Optimizing...\n",
      "New align prob=624.48\n",
      "New align prob=627.25\n",
      "New align prob=627.21\n",
      "Step 4 of 5:  Getting start-codon usage\n",
      "Step 5 of 5:  Running Glimmer3\n",
      "Starting at Wed Apr 23 13:20:23 2025\n",
      "\n",
      "Sequence file = sequence.fasta\n",
      "Number of sequences = 3\n",
      "ICM model file = training.icm\n",
      "Excluded regions file = none\n",
      "List of orfs file = none\n",
      "Input is NOT separate orfs\n",
      "Independent (non-coding) scores are used\n",
      "Circular genome = true\n",
      "Truncated orfs = false\n",
      "Minimum gene length = 110 bp\n",
      "Maximum overlap bases = 50\n",
      "Threshold score = 30\n",
      "Use first start codon = false\n",
      "Start codons = atg,gtg,ttg\n",
      "Start probs = 0.552,0.005,0.005\n",
      "Stop codons = taa,tag,tga\n",
      "GC percentage = 63.1%\n",
      "Ignore score on orfs longer than 1181\n",
      "Analyzing Sequence #1\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n",
      "Analyzing Sequence #2\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n",
      "Analyzing Sequence #3\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n"
     ]
    }
   ],
   "source": [
    "# Glimmer-from-training\n",
    "folder = \"Natronomonas\"\n",
    "portion = \"2.3\"\n",
    "process_gff(\n",
    "    gff_path=f\"Verified-genome-ALL/{folder}/verified.gff\",\n",
    "    output_path=f\"{folder}/coords-{portion}.nh\",\n",
    "    portion=portion\n",
    ")\n",
    "\n",
    "# script\n",
    "!cd Natronomonas && csh g3-from-training.csh sequence.fasta coords-2.3.nh training\n",
    "!cd Natronomonas && mkdir -p training-2.3\n",
    "!cd Natronomonas && mv training.detail training.icm training.motif training.predict training.train training.upstream training-2.3/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Saved 3.3 GFF to: Natronomonas/coords-3.3.nh\n",
      "Step 1 of 5:  Extracting training sequences\n",
      "Step 2 of 5:  Building ICM\n",
      "Step 3 of 5:  Making PWM from upstream regions\n",
      "[1]37: alignProb=672.86 Info/param=0.03 diff=6.26     \n",
      "[2]23: alignProb=664.98 Info/param=0.02 diff=21.07    \n",
      "[3]35: alignProb=692.80 Info/param=0.02 diff=28.14    \n",
      "[4]45: alignProb=732.35 Info/param=0.03 diff=17.99     \n",
      "[5]52: alignProb=746.74 Info/param=0.03 diff=21.34    \n",
      "[6]34: alignProb=657.73 Info/param=0.03 diff=32.47    \n",
      "[7]48: alignProb=642.41 Info/param=0.03 diff=8.13     \n",
      "[8]15: alignProb=695.19 Info/param=0.03 diff=20.18    \n",
      "[9]37: alignProb=756.26 Info/param=0.03 diff=18.93    \n",
      "[10]63: alignProb=755.39 Info/param=0.03 diff=20.57    \n",
      "Optimizing...\n",
      "New align prob=889.04\n",
      "New align prob=893.48\n",
      "New align prob=893.86\n",
      "New align prob=893.92\n",
      "Step 4 of 5:  Getting start-codon usage\n",
      "Step 5 of 5:  Running Glimmer3\n",
      "Starting at Wed Apr 23 13:20:40 2025\n",
      "\n",
      "Sequence file = sequence.fasta\n",
      "Number of sequences = 3\n",
      "ICM model file = training.icm\n",
      "Excluded regions file = none\n",
      "List of orfs file = none\n",
      "Input is NOT separate orfs\n",
      "Independent (non-coding) scores are used\n",
      "Circular genome = true\n",
      "Truncated orfs = false\n",
      "Minimum gene length = 110 bp\n",
      "Maximum overlap bases = 50\n",
      "Threshold score = 30\n",
      "Use first start codon = false\n",
      "Start codons = atg,gtg,ttg\n",
      "Start probs = 0.543,0.013,0.003\n",
      "Stop codons = taa,tag,tga\n",
      "GC percentage = 63.1%\n",
      "Ignore score on orfs longer than 1181\n",
      "Analyzing Sequence #1\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n",
      "Analyzing Sequence #2\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n",
      "Analyzing Sequence #3\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n"
     ]
    }
   ],
   "source": [
    "# Glimmer-from-training\n",
    "folder = \"Natronomonas\"\n",
    "portion = \"3.3\"\n",
    "process_gff(\n",
    "    gff_path=f\"Verified-genome-ALL/{folder}/verified.gff\",\n",
    "    output_path=f\"{folder}/coords-{portion}.nh\",\n",
    "    portion=portion\n",
    ")\n",
    "\n",
    "# script\n",
    "!cd Natronomonas && csh g3-from-training.csh sequence.fasta coords-3.3.nh training\n",
    "!cd Natronomonas && mkdir -p training-3.3\n",
    "!cd Natronomonas && mv training.detail training.icm training.motif training.predict training.train training.upstream training-3.3/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Roseobacter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Saved 1.3 GFF to: Roseobacter/coords-1.3.nh\n",
      "Step 1 of 5:  Extracting training sequences\n",
      "Step 2 of 5:  Building ICM\n",
      "Step 3 of 5:  Making PWM from upstream regions\n",
      "[1]36: alignProb=398.78 Info/param=0.03 diff=8.93     \n",
      "[2]58: alignProb=411.43 Info/param=0.04 diff=19.05    \n",
      "[3]12: alignProb=407.70 Info/param=0.03 diff=1.62     \n",
      "[4]22: alignProb=368.73 Info/param=0.03 diff=6.50     \n",
      "[5]21: alignProb=387.68 Info/param=0.03 diff=7.09     \n",
      "[6]30: alignProb=404.63 Info/param=0.04 diff=6.37     \n",
      "[7]13: alignProb=402.90 Info/param=0.04 diff=42.31    \n",
      "[8]35: alignProb=368.42 Info/param=0.02 diff=2.37     \n",
      "[9]32: alignProb=390.62 Info/param=0.04 diff=13.46    \n",
      "[10]48: alignProb=362.08 Info/param=0.03 diff=2.14     \n",
      "Optimizing...\n",
      "New align prob=494.90\n",
      "New align prob=506.04\n",
      "New align prob=506.30\n",
      "Step 4 of 5:  Getting start-codon usage\n",
      "Step 5 of 5:  Running Glimmer3\n",
      "Starting at Wed Apr 23 13:20:54 2025\n",
      "\n",
      "Sequence file = sequence.fasta\n",
      "Number of sequences = 5\n",
      "ICM model file = training.icm\n",
      "Excluded regions file = none\n",
      "List of orfs file = none\n",
      "Input is NOT separate orfs\n",
      "Independent (non-coding) scores are used\n",
      "Circular genome = true\n",
      "Truncated orfs = false\n",
      "Minimum gene length = 110 bp\n",
      "Maximum overlap bases = 50\n",
      "Threshold score = 30\n",
      "Use first start codon = false\n",
      "Start codons = atg,gtg,ttg\n",
      "Start probs = 0.457,0.034,0.011\n",
      "Stop codons = taa,tag,tga\n",
      "GC percentage = 58.9%\n",
      "Ignore score on orfs longer than 992\n",
      "Analyzing Sequence #1\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n",
      "Analyzing Sequence #2\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n",
      "Analyzing Sequence #3\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n",
      "Analyzing Sequence #4\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n",
      "Analyzing Sequence #5\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n"
     ]
    }
   ],
   "source": [
    "# Glimmer-from-training\n",
    "folder = \"Roseobacter\"\n",
    "portion = \"1.3\"\n",
    "process_gff(\n",
    "    gff_path=f\"Verified-genome-ALL/{folder}/verified.gff\",\n",
    "    output_path=f\"{folder}/coords-{portion}.nh\",\n",
    "    portion=portion\n",
    ")\n",
    "\n",
    "# script\n",
    "!cd Roseobacter && csh g3-from-training.csh sequence.fasta coords-1.3.nh training\n",
    "!cd Roseobacter && mkdir -p training-1.3\n",
    "!cd Roseobacter && mv training.detail training.icm training.motif training.predict training.train training.upstream training-1.3/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Saved 1.2 GFF to: Roseobacter/coords-1.2.nh\n",
      "Step 1 of 5:  Extracting training sequences\n",
      "Step 2 of 5:  Building ICM\n",
      "Step 3 of 5:  Making PWM from upstream regions\n",
      "[1]45: alignProb=508.83 Info/param=0.04 diff=21.23    \n",
      "[2]10: alignProb=483.22 Info/param=0.03 diff=58.43    \n",
      "[3]29: alignProb=587.31 Info/param=0.03 diff=19.12    \n",
      "[4]29: alignProb=493.95 Info/param=0.03 diff=3.51     \n",
      "[5]17: alignProb=468.97 Info/param=0.03 diff=3.29     \n",
      "[6]65: alignProb=586.59 Info/param=0.02 diff=9.76     \n",
      "[7]49: alignProb=490.06 Info/param=0.02 diff=8.42     \n",
      "[8]36: alignProb=489.16 Info/param=0.03 diff=4.86     \n",
      "[9]70: alignProb=541.21 Info/param=0.02 diff=2.88     \n",
      "[10]113: alignProb=577.43 Info/param=0.02 diff=1.44    \n",
      "Optimizing...\n",
      "New align prob=714.30\n",
      "New align prob=720.42\n",
      "New align prob=722.34\n",
      "New align prob=722.68\n",
      "New align prob=723.08\n",
      "New align prob=723.10\n",
      "New align prob=723.16\n",
      "Step 4 of 5:  Getting start-codon usage\n",
      "Step 5 of 5:  Running Glimmer3\n",
      "Starting at Wed Apr 23 13:21:15 2025\n",
      "\n",
      "Sequence file = sequence.fasta\n",
      "Number of sequences = 5\n",
      "ICM model file = training.icm\n",
      "Excluded regions file = none\n",
      "List of orfs file = none\n",
      "Input is NOT separate orfs\n",
      "Independent (non-coding) scores are used\n",
      "Circular genome = true\n",
      "Truncated orfs = false\n",
      "Minimum gene length = 110 bp\n",
      "Maximum overlap bases = 50\n",
      "Threshold score = 30\n",
      "Use first start codon = false\n",
      "Start codons = atg,gtg,ttg\n",
      "Start probs = 0.475,0.042,0.019\n",
      "Stop codons = taa,tag,tga\n",
      "GC percentage = 58.9%\n",
      "Ignore score on orfs longer than 992\n",
      "Analyzing Sequence #1\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n",
      "Analyzing Sequence #2\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n",
      "Analyzing Sequence #3\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n",
      "Analyzing Sequence #4\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n",
      "Analyzing Sequence #5\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n"
     ]
    }
   ],
   "source": [
    "# Glimmer-from-training\n",
    "folder = \"Roseobacter\"\n",
    "portion = \"1.2\"\n",
    "process_gff(\n",
    "    gff_path=f\"Verified-genome-ALL/{folder}/verified.gff\",\n",
    "    output_path=f\"{folder}/coords-{portion}.nh\",\n",
    "    portion=portion\n",
    ")\n",
    "\n",
    "# script\n",
    "!cd Roseobacter && csh g3-from-training.csh sequence.fasta coords-1.2.nh training\n",
    "!cd Roseobacter && mkdir -p training-1.2\n",
    "!cd Roseobacter && mv training.detail training.icm training.motif training.predict training.train training.upstream training-1.2/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Saved 2.3 GFF to: Roseobacter/coords-2.3.nh\n",
      "Step 1 of 5:  Extracting training sequences\n",
      "Step 2 of 5:  Building ICM\n",
      "Step 3 of 5:  Making PWM from upstream regions\n",
      "[1]34: alignProb=634.18 Info/param=0.02 diff=16.45    \n",
      "[2]43: alignProb=657.13 Info/param=0.02 diff=10.24    \n",
      "[3]47: alignProb=730.68 Info/param=0.03 diff=14.91    \n",
      "[4]62: alignProb=667.92 Info/param=0.03 diff=7.07     \n",
      "[5]37: alignProb=675.44 Info/param=0.03 diff=27.52    \n",
      "[6]52: alignProb=703.15 Info/param=0.02 diff=0.05     \n",
      "[7]72: alignProb=722.43 Info/param=0.03 diff=13.42    \n",
      "[8]31: alignProb=680.59 Info/param=0.03 diff=40.52    \n",
      "[9]44: alignProb=675.19 Info/param=0.03 diff=7.22     \n",
      "[10]65: alignProb=714.88 Info/param=0.03 diff=8.55     \n",
      "Optimizing...\n",
      "New align prob=980.66\n",
      "New align prob=985.42\n",
      "New align prob=987.18\n",
      "New align prob=987.27\n",
      "Step 4 of 5:  Getting start-codon usage\n",
      "Step 5 of 5:  Running Glimmer3\n",
      "Starting at Wed Apr 23 13:21:37 2025\n",
      "\n",
      "Sequence file = sequence.fasta\n",
      "Number of sequences = 5\n",
      "ICM model file = training.icm\n",
      "Excluded regions file = none\n",
      "List of orfs file = none\n",
      "Input is NOT separate orfs\n",
      "Independent (non-coding) scores are used\n",
      "Circular genome = true\n",
      "Truncated orfs = false\n",
      "Minimum gene length = 110 bp\n",
      "Maximum overlap bases = 50\n",
      "Threshold score = 30\n",
      "Use first start codon = false\n",
      "Start codons = atg,gtg,ttg\n",
      "Start probs = 0.460,0.031,0.017\n",
      "Stop codons = taa,tag,tga\n",
      "GC percentage = 58.9%\n",
      "Ignore score on orfs longer than 992\n",
      "Analyzing Sequence #1\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n",
      "Analyzing Sequence #2\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n",
      "Analyzing Sequence #3\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n",
      "Analyzing Sequence #4\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n",
      "Analyzing Sequence #5\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n"
     ]
    }
   ],
   "source": [
    "# Glimmer-from-training\n",
    "folder = \"Roseobacter\"\n",
    "portion = \"2.3\"\n",
    "process_gff(\n",
    "    gff_path=f\"Verified-genome-ALL/{folder}/verified.gff\",\n",
    "    output_path=f\"{folder}/coords-{portion}.nh\",\n",
    "    portion=portion\n",
    ")\n",
    "\n",
    "# script\n",
    "!cd Roseobacter && csh g3-from-training.csh sequence.fasta coords-2.3.nh training\n",
    "!cd Roseobacter && mkdir -p training-2.3\n",
    "!cd Roseobacter && mv training.detail training.icm training.motif training.predict training.train training.upstream training-2.3/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "✅ Saved 3.3 GFF to: Roseobacter/coords-3.3.nh\n",
      "Step 1 of 5:  Extracting training sequences\n",
      "Step 2 of 5:  Building ICM\n",
      "Step 3 of 5:  Making PWM from upstream regions\n",
      "[1]64: alignProb=1013.93 Info/param=0.02 diff=7.74     \n",
      "[2]37: alignProb=1000.99 Info/param=0.03 diff=57.75    \n",
      "[3]97: alignProb=1067.75 Info/param=0.03 diff=12.56    \n",
      "[4]9: alignProb=808.30 Info/param=0.03 diff=184.86    \n",
      "[5]74: alignProb=1142.90 Info/param=0.03 diff=11.46    \n",
      "[6]33: alignProb=979.82 Info/param=0.03 diff=1.13     \n",
      "[7]42: alignProb=1021.22 Info/param=0.03 diff=7.70     \n",
      "[8]39: alignProb=1097.34 Info/param=0.03 diff=32.51    \n",
      "[9]28: alignProb=938.39 Info/param=0.03 diff=0.71     \n",
      "[10]56: alignProb=1099.08 Info/param=0.03 diff=8.22     \n",
      "Optimizing...\n",
      "New align prob=1472.60\n",
      "New align prob=1487.75\n",
      "New align prob=1494.58\n",
      "New align prob=1494.66\n",
      "New align prob=1494.94\n",
      "Step 4 of 5:  Getting start-codon usage\n",
      "Step 5 of 5:  Running Glimmer3\n",
      "Starting at Wed Apr 23 13:22:00 2025\n",
      "\n",
      "Sequence file = sequence.fasta\n",
      "Number of sequences = 5\n",
      "ICM model file = training.icm\n",
      "Excluded regions file = none\n",
      "List of orfs file = none\n",
      "Input is NOT separate orfs\n",
      "Independent (non-coding) scores are used\n",
      "Circular genome = true\n",
      "Truncated orfs = false\n",
      "Minimum gene length = 110 bp\n",
      "Maximum overlap bases = 50\n",
      "Threshold score = 30\n",
      "Use first start codon = false\n",
      "Start codons = atg,gtg,ttg\n",
      "Start probs = 0.445,0.029,0.011\n",
      "Stop codons = taa,tag,tga\n",
      "GC percentage = 58.9%\n",
      "Ignore score on orfs longer than 992\n",
      "Analyzing Sequence #1\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n",
      "Analyzing Sequence #2\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n",
      "Analyzing Sequence #3\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n",
      "Analyzing Sequence #4\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n",
      "Analyzing Sequence #5\n",
      "Start Find_Orfs\n",
      "Start Score_Orfs\n",
      "Start Process_Events\n",
      "Start Trace_Back\n"
     ]
    }
   ],
   "source": [
    "# Glimmer-from-training\n",
    "folder = \"Roseobacter\"\n",
    "portion = \"3.3\"\n",
    "process_gff(\n",
    "    gff_path=f\"Verified-genome-ALL/{folder}/verified.gff\",\n",
    "    output_path=f\"{folder}/coords-{portion}.nh\",\n",
    "    portion=portion\n",
    ")\n",
    "\n",
    "# script\n",
    "!cd Roseobacter && csh g3-from-training.csh sequence.fasta coords-3.3.nh training\n",
    "!cd Roseobacter && mkdir -p training-3.3\n",
    "!cd Roseobacter && mv training.detail training.icm training.motif training.predict training.train training.upstream training-3.3/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### -------------------------------------------------------------------------------- GeneMarkS --------------------------------------------------------------------------------\n",
    "\n",
    "> Installation: https://genemark.bme.gatech.edu/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "os.environ[\"PATH\"] += os.pathsep + \"/home/biolab-office-1/DATALAB/GeneLM-Benchmarck/genemark_suite_linux_64/gmsuite\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# !cd Ecoli/GeneMarkS && gmsn.pl ../sequence.fasta --species Ecoli --clean\n",
    "# !cd Ecoli/GeneMarkS && gc ../sequence.fasta\n",
    "# !cd Ecoli/GeneMarkS && gmhmmp -m ../../../../genemark_suite_linux_64/gmsuite/heuristic_mod/heu_11_50.mod -f G -o genemark-hmm.gff ../sequence.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../sequence.fasta: GC% = 50.8\n"
     ]
    }
   ],
   "source": [
    "!cd Ecoli/GeneMark-HMM && gmsn.pl ../sequence.fasta --species Ecoli --clean\n",
    "!cd Ecoli/GeneMark-HMM && gc ../sequence.fasta\n",
    "!cd Ecoli/GeneMark-HMM && gmhmmp -m ../../../../genemark-2.5m/linux_64/heuristic_mod/heu_11_50.mod -f G -o genemark-hmm-2.5m.gff ../sequence.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# !cd Halobacterium/GeneMarkS && gmsn.pl ../sequence.fasta --species Halobacterium --clean\n",
    "# !cd Halobacterium/GeneMarkS && gc ../sequence.fasta\n",
    "# !cd Halobacterium/GeneMarkS && gmhmmp -m ../../../../genemark_suite_linux_64/gmsuite/heuristic_mod/heu_11_65.mod -f G -o genemark-hmm.gff ../sequence.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../sequence.fasta: GC% = 65.7\n"
     ]
    }
   ],
   "source": [
    "!cd Halobacterium/GeneMark-HMM && gmsn.pl ../sequence.fasta --species Halobacterium --clean\n",
    "!cd Halobacterium/GeneMark-HMM && gc ../sequence.fasta\n",
    "!cd Halobacterium/GeneMark-HMM && gmhmmp -m ../../../../genemark-2.5m/linux_64/heuristic_mod/heu_11_65.mod -f G -o genemark-hmm-2.5m.gff ../sequence.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# !cd Mycobacterium/GeneMarkS && gmsn.pl ../sequence.fasta --species Mycobacterium --clean\n",
    "# !cd Mycobacterium/GeneMarkS && gc ../sequence.fasta\n",
    "# !cd Mycobacterium/GeneMarkS && gmhmmp -m ../../../../genemark_suite_linux_64/gmsuite/heuristic_mod/heu_11_65.mod -f G -o genemark-hmm.gff ../sequence.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../sequence.fasta: GC% = 65.6\n"
     ]
    }
   ],
   "source": [
    "!cd Mycobacterium/GeneMark-HMM && gmsn.pl ../sequence.fasta --species Mycobacterium --clean\n",
    "!cd Mycobacterium/GeneMark-HMM && gc ../sequence.fasta\n",
    "!cd Mycobacterium/GeneMark-HMM && gmhmmp -m ../../../../genemark-2.5m/linux_64/heuristic_mod/heu_11_65.mod -f G -o genemark-hmm-2.5m.gff ../sequence.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# !cd Natronomonas/GeneMarkS && gmsn.pl ../sequence.fasta --species Natronomonas --clean\n",
    "# !cd Natronomonas/GeneMarkS && gc ../sequence.fasta\n",
    "# !cd Natronomonas/GeneMarkS && gmhmmp -m ../../../../genemark_suite_linux_64/gmsuite/heuristic_mod/heu_11_63.mod -f G -o genemark-hmm.gff ../sequence.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../sequence.fasta: GC% = 63.1\n"
     ]
    }
   ],
   "source": [
    "!cd Natronomonas/GeneMark-HMM && gmsn.pl ../sequence.fasta --species Natronomonas --clean\n",
    "!cd Natronomonas/GeneMark-HMM && gc ../sequence.fasta\n",
    "!cd Natronomonas/GeneMark-HMM && gmhmmp -m ../../../../genemark-2.5m/linux_64/heuristic_mod/heu_11_63.mod -f G -o genemark-hmm-2.5m.gff ../sequence.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# !cd Roseobacter/GeneMarkS && gmsn.pl ../sequence.fasta --species Roseobacter --clean\n",
    "# !cd Roseobacter/GeneMarkS && gc ../sequence.fasta\n",
    "# !cd Roseobacter/GeneMarkS && gmhmmp -m ../../../../genemark_suite_linux_64/gmsuite/heuristic_mod/heu_11_58.mod -f G -o genemark-hmm.gff ../sequence.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../sequence.fasta: GC% = 58.9\n"
     ]
    }
   ],
   "source": [
    "!cd Roseobacter/GeneMark-HMM && gmsn.pl ../sequence.fasta --species Roseobacter --clean\n",
    "!cd Roseobacter/GeneMark-HMM && gc ../sequence.fasta\n",
    "!cd Roseobacter/GeneMark-HMM && gmhmmp -m ../../../../genemark-2.5m/linux_64/heuristic_mod/heu_11_63.mod -f G -o genemark-hmm-2.5m.gff ../sequence.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### -------------------------------------------------------------------------------- Covert output to suitable format --------------------------------------------------------------------------------\n",
    "\n",
    "> NC_008209.1\tverified\tCDS\t3553\t4539\t.\t+\t0\tgene_id=NC008209_+_4536 start_codon=ATG type=equal\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import re\n",
    "\n",
    "def parse_genemark_lst_to_gff(lst_file_path, gff_file_path):\n",
    "    with open(lst_file_path, 'r') as infile, open(gff_file_path, 'w') as outfile:\n",
    "        outfile.write(\"##gff-version 3\\n\")\n",
    "        \n",
    "        seq_id = None\n",
    "        in_table = False\n",
    "        \n",
    "        for line in infile:\n",
    "            line = line.strip()\n",
    "            \n",
    "            # Get FASTA sequence ID\n",
    "            if line.startswith(\"FASTA definition line:\"):\n",
    "                seq_id = line.split(\":\")[1].strip()\n",
    "            \n",
    "            # Look for start of the prediction table\n",
    "            if line.startswith(\"Predicted genes\"):\n",
    "                in_table = True\n",
    "                continue\n",
    "            \n",
    "            if in_table and re.match(r\"^\\s*\\d+\\s\", line):\n",
    "                parts = re.split(r'\\s+', line)\n",
    "                if len(parts) < 6:\n",
    "                    continue\n",
    "\n",
    "                gene_id = parts[0]\n",
    "                strand = parts[1]\n",
    "                start = parts[2].replace('<', '')  # Remove '<' if present\n",
    "                end = parts[3].replace('>', '')    # Remove '>' if present\n",
    "                length = parts[4]\n",
    "                class_id = parts[5]\n",
    "\n",
    "                strand = '+' if strand == '+' else '-'\n",
    "\n",
    "                # Gene feature\n",
    "                outfile.write(f\"{seq_id}\\tGeneMark.hmm\\tgene\\t{start}\\t{end}\\t.\\t{strand}\\t.\\tID=gene{gene_id};length={length};class={class_id}\\n\")\n",
    "                # CDS feature\n",
    "                outfile.write(f\"{seq_id}\\tGeneMark.hmm\\tCDS\\t{start}\\t{end}\\t.\\t{strand}\\t0\\tID=cds{gene_id};Parent=gene{gene_id}\\n\")\n",
    "\n",
    "folders = [\"Ecoli\", \"Halobacterium\", \"Mycobacterium\", \"Natronomonas\", \"Roseobacter\"]\n",
    "glimmer_methods = {\n",
    "    \"from-scratch\": \"from-scratch.predict\",\n",
    "    \"iterated\": \"iterated.predict\",\n",
    "    # \"training-1.2\": \"training.predict\",\n",
    "    # \"training-1.3\": \"training.predict\",\n",
    "    # \"training-2.3\": \"training.predict\",\n",
    "    # \"training-3.3\": \"training.predict\"\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GeneMark-HMM: GFF saved to Ecoli/All/genemark-hmm-2.5m.gff\n",
      "GeneMark-HMM: GFF saved to Halobacterium/All/genemark-hmm-2.5m.gff\n",
      "GeneMark-HMM: GFF saved to Mycobacterium/All/genemark-hmm-2.5m.gff\n",
      "GeneMark-HMM: GFF saved to Natronomonas/All/genemark-hmm-2.5m.gff\n",
      "GeneMark-HMM: GFF saved to Roseobacter/All/genemark-hmm-2.5m.gff\n"
     ]
    }
   ],
   "source": [
    "for folder in folders:\n",
    "    # -----------------------------------------------------------------\n",
    "    # GeneMarkS\n",
    "    # -----------------------------------------------------------------\n",
    "    # input_path = os.path.join(folder, \"GeneMarkS\", \"sequence.fasta.lst\")\n",
    "    # output_root = os.path.join(folder, \"All\")\n",
    "    # output_path = os.path.join(folder, \"All\", \"genemark-s.gff\")\n",
    "    # os.makedirs(output_root, exist_ok=True)\n",
    "\n",
    "    # if not os.path.exists(input_path):\n",
    "    #     print(f\"File not found: {input_path}\")\n",
    "    #     continue\n",
    "\n",
    "    # with open(input_path, \"r\") as f:\n",
    "    #     lines = f.readlines()\n",
    "\n",
    "    # current_seq_id = None\n",
    "    # gff_lines = []\n",
    "    # for line in lines:\n",
    "    #     line = line.strip()\n",
    "    #     if line.startswith(\"FASTA definition line:\"):\n",
    "    #         match = re.search(r\"FASTA definition line:\\s*(\\S+)\", line)\n",
    "    #         if match:\n",
    "    #             current_seq_id = match.group(1)\n",
    "    #     elif re.match(r\"^\\d\", line):\n",
    "    #         parts = line.split()\n",
    "    #         if len(parts) >= 6:\n",
    "    #             gene_id = parts[0]\n",
    "    #             strand = parts[1]\n",
    "    #             left = parts[2].replace('<', '')\n",
    "    #             right = parts[3].replace('>', '')\n",
    "    #             gene_length = parts[4]\n",
    "    #             gff_line = f\"{current_seq_id}\\tGeneMarkS\\tCDS\\t{left}\\t{right}\\t.\\t{strand}\\t0\\tGeneID={gene_id};Length={gene_length}\"\n",
    "    #             gff_lines.append(gff_line)\n",
    "                \n",
    "    # with open(output_path, \"w\") as out:\n",
    "    #     out.write(\"\\n\".join(gff_lines))\n",
    "    # print(f\"\\nGeneMark: GFF written for {folder} to {output_path}\")\n",
    "    \n",
    "    # -----------------------------------------------------------------\n",
    "    # GeneMark-hmm\n",
    "    # -----------------------------------------------------------------\n",
    "    input_path = os.path.join(folder, \"GeneMark-HMM\", \"genemark-hmm-2.5m.gff\")\n",
    "    output_root = os.path.join(folder, \"All\")\n",
    "    output_path = os.path.join(output_root, \"genemark-hmm-2.5m.gff\")\n",
    "    os.makedirs(output_root, exist_ok=True)\n",
    "\n",
    "    if not os.path.exists(input_path):\n",
    "        print(f\"File not found: {input_path}\")\n",
    "    else:\n",
    "        gff_lines = []\n",
    "\n",
    "        with open(input_path, \"r\") as f:\n",
    "            for line in f:\n",
    "                line = line.strip()\n",
    "                if line.startswith(\"#\"):\n",
    "                    continue  # Skip header/comments\n",
    "                parts = line.split(\"\\t\")\n",
    "                if len(parts) == 9 and parts[2] == \"CDS\":\n",
    "                    gff_lines.append(line)\n",
    "\n",
    "        with open(output_path, \"w\") as out:\n",
    "            out.write(\"##gff-version 2\\n\")\n",
    "            out.write(\"\\n\".join(gff_lines))\n",
    "        print(f\"GeneMark-HMM: GFF saved to {output_path}\")\n",
    "    \n",
    "    # -----------------------------------------------------------------\n",
    "    # Glimmer-scratch\n",
    "    # -----------------------------------------------------------------\n",
    "    for method_folder, predict_file in glimmer_methods.items():\n",
    "        input_path = os.path.join(folder, method_folder, predict_file)\n",
    "        output_path = os.path.join(folder, \"All\", f\"glimmer-{method_folder}.gff\")\n",
    "\n",
    "        if not os.path.exists(input_path):\n",
    "            print(f\"File not found: {input_path}\")\n",
    "            continue\n",
    "\n",
    "        with open(input_path, \"r\") as f:\n",
    "            lines = f.readlines()\n",
    "\n",
    "        gff_lines = []\n",
    "        current_seq_id = None\n",
    "\n",
    "        for line in lines:\n",
    "            line = line.strip()\n",
    "            if line.startswith(\">\"):\n",
    "                current_seq_id = line[1:].split()[0]\n",
    "            elif line:\n",
    "                parts = line.split()\n",
    "                if len(parts) >= 5:\n",
    "                    gene_id = parts[0]\n",
    "                    start = int(parts[1])\n",
    "                    end = int(parts[2])\n",
    "                    strand_code = parts[3]\n",
    "                    score = parts[4]\n",
    "                    strand = '+' if '-' not in strand_code else '-'\n",
    "                    gff_line = f\"{current_seq_id}\\tGlimmer\\tCDS\\t{start}\\t{end}\\t.\\t{strand}\\t0\\tGeneID={gene_id};Score={score}\"\n",
    "                    gff_lines.append(gff_line)\n",
    "\n",
    "        os.makedirs(os.path.dirname(output_path), exist_ok=True)\n",
    "        with open(output_path, \"w\") as out:\n",
    "            out.write(\"\\n\".join(gff_lines))\n",
    "\n",
    "        print(f\"Glimer\\t: GFF written for {folder}/{method_folder} to {output_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# END"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "undefined.undefined.undefined"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
